libs/nss.c

```diff
@@ -91,57 +91,80 @@ static FILE *tcb_safe_open(const char *file, const char *name)
 	return f;
 }
 
-int _nss_tcb_getspnam_r(const char *name, struct spwd *__result_buf,
-    char *__buffer, size_t __buflen, struct spwd **__result)
+enum nss_status _nss_tcb_getspnam_r(const char *name,
+    struct spwd *__result_buf, char *__buffer, size_t __buflen, int *__errnop)
 {
 	FILE *f;
 	char *file;
 	int retval, saved_errno;
 
 	/* Disallow potentially-malicious user names */
 	if (!is_valid_username(name)) {
-		errno = ENOENT;
+		/* we don't serve an entry here */
+		*__errnop = ENOENT;
 		return NSS_STATUS_NOTFOUND;
 	}
 
-	if (asprintf(&file, TCB_FMT, name) < 0)
+	if (asprintf(&file, TCB_FMT, name) < 0) {
+		/* retry, as malloc or another resource has failed */
+		*__errnop = EAGAIN;
 		return NSS_STATUS_TRYAGAIN;
+	}
+
 	f = tcb_safe_open(file, name);
 	free(file);
-	if (!f)
-		return errno == ENOENT ? NSS_STATUS_NOTFOUND : NSS_STATUS_UNAVAIL;
+	if (!f) {
+		/* $user/shadow not existing nor readable */
+		*__errnop = ENOENT;
+		return NSS_STATUS_UNAVAIL;
+	}
```

Comment:
The original logic seems to be correct: `ENOENT` means that the shadow file doesn't exist, in that case `NSS_STATUS_NOTFOUND` (The requested entry is not available) should be returned.  If the shadow file is not available for other reasons, than `NSS_STATUS_UNAVAIL` should be returned.